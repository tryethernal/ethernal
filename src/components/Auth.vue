<template>
    <v-layout fill-height class="background">
        <v-row class="fill-height my-0">
            <v-col cols="7" class="fill-height">
                <v-row class="fill-height" align="center">
                    <v-col class="text-center">
                        <h1 class="logo">Ethernal</h1>
                        <p>
                            Ethernal is an open source block explorer for private EVM-based chains.
                        </p>
                        <p>
                            If you are new to Ethernal, you should keep the <a href="https://doc.tryethernal.com" target="_blank">doc</a> in a tab nearby!
                        </p>
                    </v-col>
                </v-row>
            </v-col>
            <v-col cols="5" class="primary fill-height">
                <v-row class="fill-height" align-self="center" align="center">
                    <v-col cols="2"></v-col>
                    <v-col cols="8">
                        <v-card v-if="signInMode" outlined>
                            <v-alert class="mb-0" dense text type="error" v-show="error">{{ error }}</v-alert>
                            <v-card-title>Sign In</v-card-title>
                            <v-card-text>
                                <v-form @submit.prevent="signIn" v-model="valid">
                                    <v-text-field
                                        :rules="[
                                            v => !!v || 'Email is required',
                                            v => /.+@.+\..+/.test(v) || 'Email must be valid',
                                        ]"
                                        required
                                        v-model="email" name="email" label="Email" type="text"></v-text-field>
                                    <v-text-field
                                        :rules="[
                                            v => !!v || 'Password is required',
                                        ]"
                                        required
                                        v-model="password" name="password" label="Password" type="password"></v-text-field>
                                    <v-card-actions class="px-0">
                                        <div style="float: left;">
                                            <small><a @click="switchMode('signup')">Sign Up</a></small><br>
                                            <small><a @click="switchMode('forgottenPwd')">Forgot My Password</a></small>
                                        </div>
                                        <v-spacer></v-spacer>
                                        <v-btn :disabled="!valid" :loading="loading" color="primary" type="submit">Sign In</v-btn>
                                    </v-card-actions>
                                </v-form>
                            </v-card-text>
                        </v-card>
                        <v-card v-else-if="signUpMode" outlined>
                            <v-alert class="mb-0" dense text type="error" v-show="error">{{ error }}</v-alert>
                            <v-card-title>Sign Up</v-card-title>
                            <v-card-text>
                                <v-form @submit.prevent="signUp" v-model="valid">
                                    <v-text-field
                                        :rules="[
                                            v => !!v || 'Email is required',
                                            v => /.+@.+\..+/.test(v) || 'Email must be valid',
                                        ]"
                                        required v-model="email" name="email" label="Email" type="text"></v-text-field>
                                    <v-text-field
                                        :rules="[
                                            v => !!v || 'Password is required',
                                        ]"
                                        required v-model="password" name="password" label="Password" type="password"></v-text-field>

                                    <v-card-actions class="px-0">
                                        <div style="float: left;">
                                            <small><a @click="switchMode('signin')">Sign In</a></small><br>
                                            <small><a @click="switchMode('forgottenPwd')">Forgot My Password</a></small>
                                        </div>
                                        <v-spacer></v-spacer>
                                        <v-btn :disabled="!valid" :loading="loading" color="primary" type="submit">Sign Up</v-btn>
                                    </v-card-actions>
                                </v-form>
                            </v-card-text>
                        </v-card>
                        <v-card v-else-if="forgottenPwdMode" outlined>
                            <v-alert class="mb-0" dense text type="error" v-show="error">{{ error }}</v-alert>
                            <v-alert class="mb-0" dense text type="success" v-show="success">{{ success }}</v-alert>
                            <v-card-title>Forgotten Password?</v-card-title>
                            <v-card-text>
                                Enter your email below and we'll send you a link to reset your password.
                                <v-form @submit.prevent="sendResetPasswordEmail" v-model="valid">
                                    <v-text-field
                                        :rules="[
                                            v => !!v || 'Email is required',
                                            v => /.+@.+\..+/.test(v) || 'Email must be valid',
                                        ]"
                                        required v-model="email" name="email" label="Email" type="text"></v-text-field>

                                    <v-card-actions class="px-0">
                                        <div style="float: left;">
                                            <small><a @click="switchMode('signin')">Sign In</a></small><br>
                                            <small><a @click="switchMode('signup')">Sign Up</a></small>
                                        </div>
                                        <v-spacer></v-spacer>
                                        <v-btn :disabled="!valid" :loading="loading" color="primary" type="submit">Submit</v-btn>
                                    </v-card-actions>
                                </v-form>
                            </v-card-text>
                        </v-card>
                        <v-card v-else-if="resetPwdMode" outlined>
                            <v-alert class="mb-0" dense text type="error" v-show="error">{{ error }}</v-alert>
                            <v-alert class="mb-0" dense text type="success" v-show="success">{{ success }}</v-alert>
                            <v-card-title>Reset Password</v-card-title>
                            <v-card-text>
                                <v-form @submit.prevent="resetPassword" v-model="valid">
                                    <v-text-field
                                        :rules="[
                                            v => !!v || 'Password is required',
                                        ]"
                                        required v-model="password" name="password" label="New Password" type="password"></v-text-field>

                                    <v-card-actions class="px-0">
                                        <div style="float: left;">
                                            <small><a @click="switchMode('signin')">Sign In</a></small><br>
                                        </div>
                                        <v-spacer></v-spacer>
                                        <v-btn :disabled="!valid" :loading="loading" color="primary" type="submit">Submit</v-btn>
                                    </v-card-actions>
                                </v-form>
                            </v-card-text>
                        </v-card>
                    </v-col>
                    <v-col cols="2"></v-col>
                </v-row>
            </v-col>
        </v-row>
    </v-layout>
</template>

<script>
export default {
    name: 'Auth',
    data: () => ({
        success: null,
        valid: false,
        error: null,
        loading: false,
        mode: 'signin',
        email: null,
        password: null
    }),
    mounted() {
        if (this.$route.query.token)
            this.mode = 'resetPwd';
    },
    methods: {
        switchMode(newMode) {
            this.error = null;
            this.email = null;
            this.password = null;
            this.success = null;
            this.mode = newMode;
        },
        signIn() {
            this.loading = true;
            this.error = null;
            this.server.signIn(this.email, this.password)
                .then(({ data: { user }}) => {
                    this.$store.dispatch('updateUser', user).then(() => document.location.href = '/transactions');
                })
                .catch(error => {
                    this.error = error.response.data;
                    this.loading = false
                });
        },
        signUp() {
            this.loading = true;
            this.error = null;
            this.server.signUp(this.email, this.password)
                .then(({ data: { user }}) => {
                    this.$store.dispatch('updateUser', user).then(() => document.location.href = '/transactions');
                })
                .catch(error => {
                    this.error = error.response.data;
                    this.loading = false
                });
        },
        sendResetPasswordEmail() {
            this.loading = true;
            this.error = null;
            this.success = null;
            this.server.sendResetPasswordEmail(this.email)
                .then(() => this.success = 'An email has been sent with a link to reset your password.')
                .catch(error => this.error = error.response.data)
                .finally(() => this.loading = false);
        },
        resetPassword() {
            this.loading = true;
            this.error = null;
            this.success = null;
            this.server.resetPassword(this.$route.query.token, this.password)
                .then(() => this.success = 'Your password has been reset successfully, you can now login.')
                .catch(error => this.error = error.response.data)
                .finally(() => this.loading = false);
        }
    },
    computed: {
        signInMode() { return this.mode == 'signin' },
        signUpMode() { return this.mode == 'signup' },
        forgottenPwdMode() { return this.mode == 'forgottenPwd' },
        resetPwdMode() { return this.mode == 'resetPwd' }
    }
}
</script>
<style scoped>
a:hover {
    text-decoration: underline;
}
</style>
