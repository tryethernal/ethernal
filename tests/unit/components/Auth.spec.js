import MockHelper from '../MockHelper';
import flushPromises from 'flush-promises'

import Auth from '@/components/Auth.vue';

const helper = new MockHelper();

describe('Auth.vue', () => {
    beforeEach(() => jest.clearAllMocks());

    it('Should load the signin screen', async () => {
        const wrapper = helper.mountFn(Auth);

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should load the signup screen', async () => {
        const wrapper = helper.mountFn(Auth);
        await wrapper.setData({ mode: 'signup' });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should load the send reset password email screen', async () => {
        const wrapper = helper.mountFn(Auth);
        await wrapper.setData({ mode: 'forgottenPwd' });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should load the reset password screen', async () => {
        const wrapper = helper.mountFn(Auth);
        await wrapper.setData({ mode: 'resetPwd' });

        expect(wrapper.html()).toMatchSnapshot();
    });
});
