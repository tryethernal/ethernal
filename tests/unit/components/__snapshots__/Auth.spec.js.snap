// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth.vue Should load the reset password screen 1`] = `
<div class="layout background fill-height">
  <div class="row fill-height my-0">
    <div class="fill-height col col-7">
      <div class="row fill-height align-center">
        <div class="text-center col">
          <h1 class="logo">Ethernal</h1>
          <p>
            Ethernal is an open source block explorer for private EVM-based chains.
          </p>
          <p>
            If you are new to Ethernal, you should keep the <a href="https://doc.tryethernal.com" target="_blank">doc</a> in a tab nearby!
          </p>
        </div>
      </div>
    </div>
    <div class="primary fill-height col col-5">
      <div align-self="center" class="row fill-height align-center">
        <div class="col col-2"></div>
        <div class="col col-8">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div role="alert" class="v-alert mb-0 v-sheet theme--dark v-alert--dense v-alert--text error--text" style="display: none;">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark error--text"></i>
                <div class="v-alert__content"></div>
              </div>
            </div>
            <div role="alert" class="v-alert mb-0 v-sheet theme--dark v-alert--dense v-alert--text success--text" style="display: none;">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-check-circle theme--dark success--text"></i>
                <div class="v-alert__content"></div>
              </div>
            </div>
            <div class="v-card__title">Reset Password</div>
            <div class="v-card__text">
              <form novalidate="novalidate" class="v-form">
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-55" class="v-label theme--light" style="left: 0px; position: absolute;">New Password</label><input required="required" name="password" id="input-55" type="password"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light">
                        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-card__actions px-0">
                  <div style="float: left;"><small><a>Sign In</a></small><br></div>
                  <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">Submit</span></button>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div class="col col-2"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Auth.vue Should load the send reset password email screen 1`] = `
<div class="layout background fill-height">
  <div class="row fill-height my-0">
    <div class="fill-height col col-7">
      <div class="row fill-height align-center">
        <div class="text-center col">
          <h1 class="logo">Ethernal</h1>
          <p>
            Ethernal is an open source block explorer for private EVM-based chains.
          </p>
          <p>
            If you are new to Ethernal, you should keep the <a href="https://doc.tryethernal.com" target="_blank">doc</a> in a tab nearby!
          </p>
        </div>
      </div>
    </div>
    <div class="primary fill-height col col-5">
      <div align-self="center" class="row fill-height align-center">
        <div class="col col-2"></div>
        <div class="col col-8">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div role="alert" class="v-alert mb-0 v-sheet theme--dark v-alert--dense v-alert--text error--text" style="display: none;">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark error--text"></i>
                <div class="v-alert__content"></div>
              </div>
            </div>
            <div role="alert" class="v-alert mb-0 v-sheet theme--dark v-alert--dense v-alert--text success--text" style="display: none;">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-check-circle theme--dark success--text"></i>
                <div class="v-alert__content"></div>
              </div>
            </div>
            <div class="v-card__title">Forgotten Password?</div>
            <div class="v-card__text">
              Enter your email below and we'll send you a link to reset your password.
              <form novalidate="novalidate" class="v-form">
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-38" class="v-label theme--light" style="left: 0px; position: absolute;">Email</label><input required="required" name="email" id="input-38" type="text"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light">
                        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-card__actions px-0">
                  <div style="float: left;"><small><a>Sign In</a></small><br> <small><a>Sign Up</a></small></div>
                  <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">Submit</span></button>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div class="col col-2"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Auth.vue Should load the signin screen 1`] = `
<div class="layout background fill-height">
  <div class="row fill-height my-0">
    <div class="fill-height col col-7">
      <div class="row fill-height align-center">
        <div class="text-center col">
          <h1 class="logo">Ethernal</h1>
          <p>
            Ethernal is an open source block explorer for private EVM-based chains.
          </p>
          <p>
            If you are new to Ethernal, you should keep the <a href="https://doc.tryethernal.com" target="_blank">doc</a> in a tab nearby!
          </p>
        </div>
      </div>
    </div>
    <div class="primary fill-height col col-5">
      <div align-self="center" class="row fill-height align-center">
        <div class="col col-2"></div>
        <div class="col col-8">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div role="alert" class="v-alert mb-0 v-sheet theme--dark v-alert--dense v-alert--text error--text" style="display: none;">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark error--text"></i>
                <div class="v-alert__content"></div>
              </div>
            </div>
            <div class="v-card__title">Sign In</div>
            <div class="v-card__text">
              <form novalidate="novalidate" class="v-form">
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-8" class="v-label theme--light" style="left: 0px; position: absolute;">Email</label><input required="required" name="email" id="input-8" type="text"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light">
                        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-11" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input required="required" name="password" id="input-11" type="password"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light">
                        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-card__actions px-0">
                  <div style="float: left;"><small><a>Sign Up</a></small><br> <small><a>Forgot My Password</a></small></div>
                  <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">Sign In</span></button>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div class="col col-2"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Auth.vue Should load the signup screen 1`] = `
<div class="layout background fill-height">
  <div class="row fill-height my-0">
    <div class="fill-height col col-7">
      <div class="row fill-height align-center">
        <div class="text-center col">
          <h1 class="logo">Ethernal</h1>
          <p>
            Ethernal is an open source block explorer for private EVM-based chains.
          </p>
          <p>
            If you are new to Ethernal, you should keep the <a href="https://doc.tryethernal.com" target="_blank">doc</a> in a tab nearby!
          </p>
        </div>
      </div>
    </div>
    <div class="primary fill-height col col-5">
      <div align-self="center" class="row fill-height align-center">
        <div class="col col-2"></div>
        <div class="col col-8">
          <div class="v-card v-sheet v-sheet--outlined theme--light">
            <div role="alert" class="v-alert mb-0 v-sheet theme--dark v-alert--dense v-alert--text error--text" style="display: none;">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark error--text"></i>
                <div class="v-alert__content"></div>
              </div>
            </div>
            <div class="v-card__title">Sign Up</div>
            <div class="v-card__text">
              <form novalidate="novalidate" class="v-form">
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-23" class="v-label theme--light" style="left: 0px; position: absolute;">Email</label><input required="required" name="email" id="input-23" type="text"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light">
                        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-input theme--light v-text-field">
                  <div class="v-input__control">
                    <div class="v-input__slot">
                      <div class="v-text-field__slot"><label for="input-26" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input required="required" name="password" id="input-26" type="password"></div>
                    </div>
                    <div class="v-text-field__details">
                      <div class="v-messages theme--light">
                        <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="v-card__actions px-0">
                  <div style="float: left;"><small><a>Sign In</a></small><br> <small><a>Forgot My Password</a></small></div>
                  <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">Sign Up</span></button>
                </div>
              </form>
            </div>
          </div>
        </div>
        <div class="col col-2"></div>
      </div>
    </div>
  </div>
</div>
`;
